### DISCUSSION TITLE
How to cancel autocomplete prompt?

### DISCUSSION THREAD

**Discussion by thelibrarian** [zed-industries/zed#11474] [Feature Requests]
Hi,

When Zed offers an autocomplete in a pop-up (i.e. not the Copilot suggestion), is there a command or hotkey to dismiss this? I've found that there are times when the suggestions are irrelevant, spurious, or simply quicker to type the keyword than to select it from the selection box.

A concrete example: when I fully type the `end` keyword in Ruby, the first autocomplete option is `Encoding`, and the `end` autocomplete is option 5 or 6, even though I've already typed the full word.

The only way that I have so far found to dismiss this pop-up is to press `Esc`, but this also takes me out of Insert Mode. Is there another command/hotkey that would dismiss this popup without also exiting insert mode?

**Comment by baldwindavid** [comment]
> A concrete example: when I fully type the end keyword in Ruby, the first autocomplete option is Encoding, and the end autocomplete is option 5 or 6, even though I've already typed the full word.

Cases like this where I want to press `enter` were very problematic. This happens because of these keymap defaults...

```json
  {
    "context": "Editor && showing_completions",
    "bindings": {
      "enter": "editor::ConfirmCompletion",
      "tab": "editor::ConfirmCompletion"
    }
  },
```

To each their own in terms of confirming completion, but I definitely don't want _both_ `enter` and `tab` doing so and added a keybinding to override for `enter`.

```json
  {
    "context": "Editor && showing_completions",
    "bindings": {
      "enter": "editor::Newline"
    }
  }
```

> there are times when the suggestions are irrelevant, spurious, or simply quicker to type the keyword than to select it from the selection box.

I haven't really had a need to solve for canceling the completion otherwise as I can just keep typing on the same line and ignoring the completion suggestions. There's _probably_ a way to override the `escape` keymap though I don't see an `editor::CancelCompletion` and `menu::Cancel` seemed to changed to normal mode with a quick test. I'm thinking that even if the completion is canceled it's just going to show up upon the next character typed, but there might be more nuance to it.

**Reply by thelibrarian** [reply]
Thank you, this fixed my problem - removing the "enter" binding from the confirm completion and adding the newline binding now effectively does what I need. When I type enter, it now goes to the next line as I want, and effectively cancels the completion, which is actually better then what I requested - no extra key press needed to cancel the completion.

**Reply by willdhorn** [reply]
> There's probably a way to override the escape keymap though I don't see an editor::CancelCompletion and menu::Cancel seemed to changed to normal mode with a quick test

You can use `editor::Cancel` instead of `menu::Cancel` to hide the completions without exiting insert mode. So if you still wanted to use escape as well you could do
```json
{
	"context": "Editor && showing_completions",
	"bindings": {
	  "enter": "editor::Newline",
	  "escape": "editor::Cancel"
	}
}
```

**Reply by ches** [reply]
Just to add search engine juice and say thanks for @baldwindavid and @willdhorn's answers which are both helpful:

> I haven't really had a need to solve for canceling the completion otherwise as I can just keep typing on the same line and ignoring the completion suggestions.

We may be a dying breed in the age of AI edit prediction, but for snippets users it comes up constantly: you are entering a snippet and want to Tab through its placeholders, but other completions keep popping up and stealing your Tab press.

Without @willdhorn's suggestion, <kbd>Escape</kbd> will exit insert mode making snippets practically unusable for Vim mode. That contextual `editor::Cancel` binding gives a workable alternative to turning off as-you-type completions altogether. 👍🏼

**Comment by bminde** [comment]
`shift-enter` gives you a new line without accepting.

**Comment by edhemphill** [comment]
I can't seem to disable tab as an option for code completion using this pattern. I have this slapped into keymap.json:

```
// from the command palette.
[
  {
    "context": "Editor && showing_completions",
    "bindings": {
      "enter": "editor::Newline",
      "tab": "editor::Tab",
      "shift-enter": "editor::ConfirmCompletion"
    }
  },
  {
    "context": "Editor",
    "bindings": {
 ...
```

I must be doing something wrong... perhaps things have changed since a few months ago?

**Reply by numist** [reply]
A bit of looking around suggests that `showing_completions` is now called `inline_completion`? That seemed to work for me, at least.

**Reply by dibenzepin** [reply]
`showing_completions` worked for me...at least in the context of rebinding `Tab`.

this is my current keymap:
```json
[
  {
    "context": "Editor && showing_completions",
    "bindings": {
      "tab": "editor::ContextMenuNext",
      "shift-tab": "editor::ContextMenuPrevious"
    }
  }
]
```
